### PR TITLE
bpo-29851: Have importlib.reload() raise ImportError if the module's spec is not found

### PR DESCRIPTION
**Pull request by plusminushalf**
if _bootstrap._find_spec() return None, an ImportError is raised
stating: module could not be found in the path specified

### CODE REVIEW

**Comment by DimitrisJim** [comment]
Hi @garvitdelhi, I think this might also require tests to verify the change.

File: Lib/importlib/__init__.py [hunk 1]
```diff
@@ -164,6 +164,9 @@ def reload(module):
             pkgpath = None
         target = module
         spec = module.__spec__ = _bootstrap._find_spec(name, pkgpath, target)
+        if spec is None:
+            msg = "module {} not found in the path specified"
```
Comment (brettcannon) [inline Lib/importlib/__init__.py]: I would change this message to `"spec not found for the module {!r}"`.

**Review by brettcannon — CHANGES_REQUESTED** [review]
Thanks for the patch, but tests are needed for this.

File: Lib/importlib/__init__.py [hunk 2]
```diff
@@ -164,6 +164,9 @@ def reload(module):
...
+            raise ImportError(msg.format(name), name=name)
```
Comment (brettcannon) [inline Lib/importlib/__init__.py]: As mentioned in the issue, this should probably be `ModuleNotFoundError`.

**Comment by plusminushalf** [comment]
Hi, I am a little busy and hence will by writing tests by the weekend.

**Comment by plusminushalf** [comment]
Hi, I tried writing tests but I am facing a problem.

I tried this:

```
moduleName = 'testmodule'
path = os.getcwd() + '/testmodule.py'
spec = importlib.util.spec_from_file_location(moduleName, path)
module_test = importlib.util.module_from_spec(spec)
spec.loader.exec_module(module_test)
sys.modules[moduleName] = module_test
with self.assertRaises(ModuleNotFoundError):
    importlib.reload(sys.modules[moduleName])
```

The problem here is I have testmodule.py within the same directory i.e Lib/test/test_importlib/extension/ but since tests are run in a sandbox environment I don't have access to above directory. Also I can't load the file itself since it has a relative import 'abc' and I get an error saying: "ImportError: attempted relative import with no known parent package"

**Comment by ncoghlan** [comment]
Hi @garvitdelhi, for cases like this, it's worth looking through the existing test cases to see if there are some existing examples you can copy, as we tend to have lots of odd helpers in the test suite that let us do things that we expressly tell end users not to do (since we're deliberately inducing error states so we can check for expected exceptions).

For this PR, it turns out the most relevant examples are in `tests.test_importlib.test.api`: https://github.com/python/cpython/blob/master/Lib/test/test_importlib/test_api.py#L198

As with the examples there, the simplest way to get a module that can't be reloaded is to take a normal module and corrupt its metadata, as in something like:

    >>> import types
    >>> import importlib
    >>> importlib.reload(types)
    <module 'types' from '/usr/lib64/python3.5/types.py'>
    >>> import sys
    >>> types.__name__ = "not_a_module"
    >>> types.__spec__ = None
    >>> sys.modules[types.__name__] = types
    >>> importlib.reload(types)
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/usr/lib64/python3.5/importlib/__init__.py", line 166, in reload
        _bootstrap._exec(spec, module)
      File "<frozen importlib._bootstrap>", line 607, in _exec
    AttributeError: 'NoneType' object has no attribute 'name'

File: Lib/importlib/__init__.py [hunk 3]
```diff
@@ -164,6 +164,9 @@ def reload(module):
             pkgpath = None
         target = module
         spec = module.__spec__ = _bootstrap._find_spec(name, pkgpath, target)
+        if spec is None:
+            msg = "spec not found for the module {!r}"
+            raise ModuleNotFoundError(msg.format(name), name=name)
```
Comment (berkerpeksag) [inline Lib/importlib/__init__.py]: You can use an f-string here:

```py
raise ModuleNotFoundError(f"spec not found for the module {name!r}", name=name)
```

File: Lib/test/test_importlib/extension/test_loader.py [hunk 4]
```diff
@@ -145,6 +146,15 @@ def test_reload(self):
             importlib.reload(module)
             self.assertIs(ex_class, module.Example)
 
+    def test_reload_faliure(self):
+        '''Test that reload throws ModuleNotFounderror if reloading'''
```
Comment (berkerpeksag) [inline Lib/test/test_importlib/extension/test_loader.py]: `"""..."""` supports multiline comments so you can just do:

```py
'''Test that reload throws ModuleNotFounderror if reloading
a module whose spec cannot be found'''
```

I'll let Brett or Nick comment on the actual wording of docstring.

File: Lib/test/test_importlib/extension/test_loader.py [hunk 5]
```diff
@@ -145,6 +146,15 @@ def test_reload(self):
...
+        '''a module whose spec cannot be found'''
+        datetime.__name__ = 'destroyed_datetime'
```
Comment (berkerpeksag) [inline Lib/test/test_importlib/extension/test_loader.py]: We probably shouldn't directly mess with a stdlib module and `sys.modules` and use a helper like Nick said. Untested, but I think we need to add something like the following snippet to `Lib/test/test_importlib/test_api.py`:

```py
with support.CleanImport('types'):
    module = self.init.import_module('types')
    module.__name__ = 'broken_types'
    module.__spec__ = None
    sys.modules[module.__name__] = module  # TODO: Do we need to cleanup sys.modules?
    with self.assertRaises(ModuleNotFoundError):
        self.init.reload(module)
```

**Review by berkerpeksag — CHANGES_REQUESTED** [review]
Please add an entry to `Misc/NEWS` (also include "Patch by Your Name.") Thanks!

File: Lib/test/test_importlib/extension/test_loader.py (hunk 4 above)
Comment (brettcannon) [reply Lib/test/test_importlib/extension/test_loader.py]: Actually the whole module is a mess when it comes to docstrings.

Use a `#` comment instead of a docstring, and it can take more than a single line. Also make sure to end it in a period.

File: Lib/test/test_importlib/extension/test_loader.py (hunk 5 above)
Comment (brettcannon) [reply Lib/test/test_importlib/extension/test_loader.py]: All the helpers for importlib-related tests can be found in https://github.com/python/cpython/blob/master/Lib/test/test_importlib/util.py. In this instance you probably want both uncache() and temp_module().

**Comment by plusminushalf** [comment]
Hey sorry for the delay was busy in relocating, back to contributions now :)

File: Lib/test/test_importlib/extension/test_loader.py [hunk 6]
```diff
@@ -145,6 +145,21 @@ def test_reload(self):
             importlib.reload(module)
             self.assertIs(ex_class, module.Example)
 
+    def test_reload_faliure(self):
```
Comment (brettcannon) [inline Lib/test/test_importlib/extension/test_loader.py]: Tests for reloading should be in [test.test_importlib.test_api](https://github.com/python/cpython/blob/master/Lib/test/test_importlib/test_api.py#L198).

**Review by brettcannon — CHANGES_REQUESTED** [review]
A versionchanged note should be added to the docs on top of the relocation of the tests. And don't forget to add yourself to Misc/ACKS.

File: .gitignore [hunk 7]
```diff
@@ -99,3 +99,4 @@ Tools/ssl/amd64
 Tools/ssl/win32
 .vs/
 .vscode/
+.idea/
```
Comment (serhiy-storchaka) [inline .gitignore]: This change doesn't look related.

Comment (plusminushalf) [reply .gitignore]: will make a different issue for this removing this as of now

File: Lib/test/test_importlib/test_api.py [hunk 8]
```diff
@@ -361,6 +361,20 @@ def test_reload_submodule(self):
             reloaded = self.init.reload(ham)
             self.assertIs(reloaded, ham)
 
+    def test_reload_faliure(self):
+        #Test that reload throws ModuleNotFounderror if reloading
+        #a module whose spec cannot be found.
+        name = 'spam'
+        subname = 'ham'
+        with test_util.temp_module(name, pkg=True) as pkg_dir:
+            fullname, _ = test_util.submodule(name, subname, pkg_dir)
```
Comment (brettcannon) [inline Lib/test/test_importlib/test_api.py]: Is there a specific reason to test a submodule?

File: Lib/test/test_importlib/test_api.py [hunk 9]
```diff
@@ -361,6 +361,20 @@ def test_reload_submodule(self):
...
+            module = self.init.import_module(fullname)
+            module.__spec__ = None
+            module.__name__ = 'destroyed_spam'
+            sys.modules[module.__name__] = module
```
Comment (brettcannon) [inline Lib/test/test_importlib/test_api.py]: Would the test still work if you simply set `module.__spec__ = None` and left the name alone and didn't insert a new module? Currently you end up leaving a dead module in sys.modules under the name `destroyed_spam`.

**Review by brettcannon — CHANGES_REQUESTED** [review]
This is also going to require a versionchanged mention in the docs.

File: .gitignore (hunk 7 above)
Comment (berkerpeksag) [reply .gitignore]: Note that you can use a global .gitignore file to ignore these IDE specific files. We've rejected a few PRs because the global .gitignore solution is the preferred way to do it.

Comment (plusminushalf) [reply .gitignore]: Thank you for pointing this out almost forgot about global git ignores.

File: Lib/test/test_importlib/test_api.py (hunk 9 above)
Comment (plusminushalf) [reply Lib/test/test_importlib/test_api.py]: Not changing the name won't work. If I don't modify the name, the method _find_spec inside importlib/_bootstrap.py tries to find spec from three finder from sys.meta_path:  _frozen_importlib.BuiltinImporter, _frozen_importlib.FrozenImporter and _frozen_importlib_external.PathFinder respectively. But since I have not changed the name and the module by that filename still exists PathFinder can find the spec (reload the spec) from the file. Hence I do need to change the name and put it explicitly in sys.modules

**Comment by brettcannon** [comment]
@garvitdelhi when you think you have addressed all the requested changes, could you remove the "[WIP]" part of the title that I added and mention me in a comment so this PR shows up as a GitHub notification for me?

**Comment by plusminushalf** [comment]
@brettcannon done :)

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
I made some of my own changes to the PR so I'm waiting to make sure I didn't break anything. Once Travis passes I will commit.

**Comment by brettcannon** [comment]
Thanks for PR, @garvitdelhi and being patient with the whole process!

**Comment by VKTB** [comment]
I know that this contribution is from a while ago but I am dynamically loading my own modules and get a `ModuleNotFoundError: spec not found for the module 'a'` error message when I call `importlib.reload(sys.modules['a'])`. 

I am loading the module using this logic::
```Python
import importlib.util

module_name = 'a'
path = os.getcwd() + '/a.py'
spec = importlib.util.spec_from_file_location(module_name, path)
module_a = importlib.util.module_from_spec(spec)
spec.loader.exec_module(module_a)
sys.modules[module_name ] = module_a
```

When I inspect the module in debugging mode, I can see that it has its own `ModuleSpec` (see screenshot below), so I do not quite understand why it is raising the `ImportError` that was introduced as part of this PR. I also tried `module_a.__spec__ = spec` before adding the module to `sys.modules` just to be safe that spec is added to the module, but no luck. Any ideas?

image

**Comment by brettcannon** [comment]
@VKTB this isn't the best place to be asking this. If you're after help you can try asking on python-list, else if you think it's a bug then please open one at bugs.python.org.
